docs/proposals/timeseries-partitioning-in-compactor.md

```diff
@@ -0,0 +1,267 @@
+
+`partitionId = Hash(timeseries label) % number of partition`
+
+The number of partition will be determined automatically, via a configured `multiplier`. This `multiplier` factor allows us to group just a subset of the blocks together to achieve the same deduplication factor as having all the blocks. Using a `multiplier` of 2 as an example, we can do grouping for partition of 2,  4 and 8. We’ll build on the actual number of partition determination in a later section.
+
+### Determining overlapping blocks
+
+In order to reduce the amount of time spent downloading blocks, and iterating through the index to filter out unrelated timeseries, we can do smart grouping of the blocks.
+
+![Modulo Partitioning](/images/proposals/timeseries-partitioning-in-compactor-modulo-partition.png)
+
+Given that we are always multiplying the number of partition by the `multiplier` factor, we can deduce from the modulo which partition could contain overlapping result
+
+```
+Given a hash N, if N % 8 == 7, then N % 4 must be 3
+Given a hash N, if N % 8 == 3, then N % 4 must be 3
+Given a hash N, if N % 8 == 4, then N % 4 must be 0
+Given a hash N, if N % 8 == 0, then N % 4 must be 0
+```
+
+Hence it is safe to group blocks with `N % 8 == 7` with `N % 8 == 3 `together with `N % 4 == 3` together, and we are sure that other blocks won’t contain the same timeseries. We also know that if `N % 8 == 0`, then we don’t need to download blocks where `N % 4 == 1` or `N % 4 == 2
+`
+Given partition count and partition id, we can immediately find out which blocks are required. Using the above modulo example, we get the following partitiong mapping.
+
+![Partition](/images/proposals/timeseries-partitioning-in-compactor-partitions.png)
+### Planning the compaction
+
+The shuffle sharding compactor introduced additional logic to group blocks by distinct time intervals. It can also sum up the sizes of all indices to determine how many shards are required in total. Using the above example again, and assuming that each block has an index of 30GB, the sum is 30GB * 14 = 420GB, which needs to be at least 7, since maximum index size is 64GB. Using the `multiplier` factor, it will be rounded up to 8.
+
+Now the planner knows the resulting compaction will have 8 partitions, it can start planning out which groups of blocks can go into a single compaction group. Given that we need 8 partitions in total, the planner will go through the process above to find out what blocks are necessary. Using the above example again, but we have distinct time intervals, T1, T2, and T3. T1 has 2 partitions, T2 has 4 partitions, and T3 has 8 partitions, and we want to produce T1-T3 blocks
+![Grouping](/images/proposals/timeseries-partitioning-in-compactor-grouping.png)
+```
+Compaction Group 1-8
```

Comment:
Right now for compaction partitioning, we split 8 jobs to 8 compactors to pick up and run the compaction. This is okay for us but I am not sure if all Cortex users value compaction speed over resources usage.
Probably we can mention or have another mode for 1 compaction job which creates 8 blocks the same time?

Reply:
If compaction is delayed, the read path needs more resources (store-gateways, queriers, etc). So I would say yes, most users want speedy compaction to avoid spending on resources on the read path.

Reply:
I agree speed is important. But resources should be also taken into consideration. Let's say a block has 16 shards so in this case we need to download it and compact it 16 times.
If a single compaction can create 8 blocks, compare to 8 compactions to generate 8 blocks. Although the latter is faster, single compaction should still be better in terms of total CPU time. Since no block and index needs to be download or verified multiple times.